spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala

```diff
@@ -1124,6 +1129,65 @@ class DeltaAnalysis(session: SparkSession)
     DeltaViewHelper.stripTempViewForMerge(plan, conf)
   }
 
+  /**
+   * Recursively casts map data types in case the key/value type differs.
+   */
+  private def addCastsToMaps(
+      tableName: String,
+      parent: NamedExpression,
+      sourceMapType: MapType,
+      targetMapType: MapType,
+      allowTypeWidening: Boolean): Expression = {
+
+    val transformedKeys =
+      if (sourceMapType.keyType != targetMapType.keyType) {
+        // Create a transformation for the keys
+        ArrayTransform(MapKeys(parent), {
+          val key = NamedLambdaVariable(
+                  "key", sourceMapType.keyType, nullable = false)
+
+          val keyAttr = AttributeReference(
+              "key", targetMapType.keyType, nullable = false)()
+
+          val castedKey =
+              addCastToColumn(
+                key,
+                keyAttr,
```

Comment:
Welcome back Johanl-db.
I can address the style issues after work this evening.
The key.toAttribute is causing the maxIteration error I wrote about earlier.

Reply:
Ah right, that's probably because the data type isn't correct with `key.toAttribute` - it's the source type where it should be the target type - and that trips the analyzer